### PR TITLE
feat: Update encoding and decoding and simulation interfaces to use alloy instead of ethers. 

### PR DESCRIPTION
**Pull request by dianacarvalho1**
Update encoding and decoding to use alloy instead of ethers
Updated this in adapter_contract.rs, tycho_simulation_contract.rs and erc20_token.rs Delete unnecessary abi handling and respective files Unify addresses as Address
In EVMPoolState change U256 to be from alloy instead of ethers.
Replace ethers types with alloy in simulation interfaces
Replace ethers types in block hash
Unify all the alloy imports and rename rU256 and rAddress

TODO:
- Token abstraction 
- Alloy should be optional
- Change inner simulation code to not use ethers - I think this is a looot and I propose not to do it now 😕

### CODE REVIEW

File: src/evm/protocol/u256_num.rs [hunk 1]
```diff
@@ -110,6 +110,32 @@ pub fn convert_ethers_to_alloy(ethers_u256: EthersU256) -> U256 {
     ethers_u256.to_big_endian(&mut bytes); // Fill the buffer with big-endian bytes
     U256::from_be_bytes(bytes) // Convert to Alloy U256
 }
+pub fn convert_alloy_to_ethers(alloy_value: U256) -> EthersU256 {
```
Comment (dianacarvalho1) [inline src/evm/protocol/u256_num.rs:113]: these two functions are temporary so I could divide this in several parts

File: src/evm/protocol/vm/state_builder.rs [hunk 2]
```diff
@@ -425,29 +422,11 @@ impl EVMPoolStateBuilder {
...
-                )
-            })?
-            .into_iter()
-            .next()
-            .ok_or_else(|| {
-                SimulationError::FatalError(
-                    "Failed to get address from call: Couldn't retrieve address from simulation for stateless contracts".into(),
-                )
-            })?;
+        let address: Address =Address::abi_decode(&sim_result.result, true).map_err(|e| {
```
Comment (tamaralipows) [inline src/evm/protocol/vm/state_builder.rs:425]: Are you nightly formatting? Spacing looks weird

Comment (dianacarvalho1) [reply src/evm/protocol/vm/state_builder.rs:425]: oh yes.. I did 😕 weird

File: src/evm/protocol/vm/tycho_simulation_contract.rs [hunk 3]
```diff
@@ -104,95 +88,43 @@ where
...
+        call_data
     }
 
     #[allow(clippy::too_many_arguments)]
     pub fn call(
         &self,
-        fname: &str,
-        args: Vec<Token>,
+        selector: &str,
+        args: impl SolValue + std::fmt::Debug,
```
Comment (tamaralipows) [inline src/evm/protocol/vm/tycho_simulation_contract.rs:118]: Interesting, never saw `impl` used before. That's great

File: src/evm/protocol/vm/utils.rs [hunk 4]
```diff
@@ -368,6 +350,22 @@ pub fn hexstring_to_vec(hexstring: &str) -> Result<Vec<u8>, SimulationError> {
     Ok(bytes)
 }
 
+pub fn string_to_bytes32(pool_id: &str) -> Result<[u8; 32], SimulationError> {
+    let pool_id_no_prefix =
+        if let Some(stripped) = pool_id.strip_prefix("0x") { stripped } else { pool_id };
+    let bytes = hex::decode(pool_id_no_prefix)
```
Comment (tamaralipows) [inline src/evm/protocol/vm/utils.rs:356]: `hex::decode` automatically pads to 32 bytes? didn't know

File: src/evm/simulation.rs [hunk 5]
```diff
@@ -646,31 +629,41 @@ mod tests {
...
+        let path = vec![usdc_addr, weth_addr];
+
+        let encoded = {
+            let args = (amount_in, path);
+            let mut hasher = Keccak256::new();
+            hasher.update(selector.as_bytes());
+            let selector_bytes = &hasher.finalize()[..4];
+            let mut data = selector_bytes.to_vec();
+            let mut encoded_args = args.abi_encode();
+            // Remove extra prefix if present (32 bytes for dynamic data)
```
Comment (tamaralipows) [inline src/evm/simulation.rs:647]: What do you mean by extra prefix? Can you clarify in the comment?

**Review by tamaralipows — APPROVED** [review]
Honestly just shocked about how clean everything is now. Very minor comments/questions. Thank you!

File: src/evm/protocol/vm/adapter_contract.rs [hunk 6]
```diff
@@ -47,197 +55,144 @@ where
...
-                res.return_value[0]
-            )))
-            }?;
+        let args = (string_to_bytes32(&pair_id)?, sell_token, buy_token, is_buy, amount);
+        let selector = "swap(bytes32,address,address,uint8,uint256)";
+
+        let res = self.call(selector, args, block, None, overwrites, None, U256::from(0u64))?;
+
+        let decoded: SwapReturn = SwapReturn::abi_decode(&res.return_value, true).map_err(|e| {
+            SimulationError::FatalError(format!("Failed to decode swap return value: {:?}", e))
```
Comment (louise-poole) [inline src/evm/protocol/vm/adapter_contract.rs:97]: Same here - why are we dropping the error details?

File: src/evm/protocol/vm/adapter_contract.rs [hunk 7]
```diff
@@ -47,197 +55,144 @@ where
 {
     pub fn price(
         &self,
-        pair_id: Vec<u8>,
+        pair_id: String,
```
Comment (louise-poole) [inline src/evm/protocol/vm/adapter_contract.rs:58]: This gets immediately converted to a reference and used only as a reference. Why not pass it as a reference here already and avoid needing to clone the pair_id at the call site?

File: src/evm/protocol/vm/adapter_contract.rs [hunk 8]
```diff
@@ -47,197 +55,144 @@ where
...
+
+        let price = self.calculate_price(decoded)?;
         Ok(price)
     }
 
     #[allow(clippy::too_many_arguments)]
     pub fn swap(
         &self,
-        pair_id: Vec<u8>,
+        pair_id: String,
```
Comment (louise-poole) [inline src/evm/protocol/vm/adapter_contract.rs:83]: Same here

File: src/evm/protocol/vm/state.rs [hunk 9]
```diff
@@ -748,21 +741,21 @@ mod tests {
...
+        assert_eq!(dai_limit, U256::from_str("100279494253364362835").unwrap());
 
         // let bal_limit = pool_state
         //     .get_sell_amount_limit(
         //         vec![pool_state.tokens[1], pool_state.tokens[0]],
         //         Some(overwrites),
         //     )
         //     .unwrap();
-        // assert_eq!(bal_limit, U256::from_dec_str("13997408640689987484").unwrap());
+        // assert_eq!(bal_limit, U256::from_str("13997408640689987484").unwrap());
```
Comment (louise-poole) [inline src/evm/protocol/vm/state.rs:758]: Do you know why is this is commented out?

File: src/evm/tycho_models.rs [hunk 10]
```diff
@@ -14,21 +12,21 @@ use crate::serde_helpers::{hex_bytes, hex_bytes_option};
 use super::engine_db::simulation_db::BlockHeader;
 
 // TODO move this to utils
-fn bytes_to_ru256(bytes: Bytes) -> rU256 {
+fn bytes_to_ru256(bytes: Bytes) -> U256 {
```
Comment (louise-poole) [inline src/evm/tycho_models.rs:15]:
```suggestion
fn bytes_to_u256(bytes: Bytes) -> U256 {
```

File: src/evm/tycho_models.rs [hunk 11]
```diff
@@ -14,21 +12,21 @@ use crate::serde_helpers::{hex_bytes, hex_bytes_option};
...
     let start = 32 - bytes.len().min(32);
     padded_bytes[start..].copy_from_slice(&bytes[bytes.len().saturating_sub(32)..]);
 
-    // Convert the padded byte array into rU256
-    rU256::from_be_slice(&padded_bytes)
+    // Convert the padded byte array into U256
+    U256::from_be_slice(&padded_bytes)
 }
 
 fn map_slots_to_ru256(
```
Comment (louise-poole) [inline src/evm/tycho_models.rs:27]:
```suggestion
fn map_slots_to_u256(
```

**Review by louise-poole — COMMENTED** [review]
Just a couple suggestions/questions

File: src/evm/protocol/vm/adapter_contract.rs (hunk 6 above)
Comment (dianacarvalho1) [reply src/evm/protocol/vm/adapter_contract.rs:97]: you right! will add it back!

File: src/evm/protocol/vm/state.rs (hunk 9 above)
Comment (dianacarvalho1) [reply src/evm/protocol/vm/state.rs:758]: good question! I think it was a mistake! it passes anyway! bringing it back! good catch! Thank you 🙏🏼

**Pull request merged**